tests/ui/run.ts

```diff
@@ -3,59 +3,134 @@
 
     // Insert beer
-    await t.click('.e1-food').click('.e1._1f37b');
-
-    // Ensure both have been inserted
-    const getChildNodeCount = await ClientFunction(() => {
-        return document.querySelector('div.compose').childNodes.length;
-    });
-    await t.expect(await getChildNodeCount()).eql(2);
-
-    const firstEmoji = await Selector('div.compose img').nth(0)();
-    await t.expect(firstEmoji.tagName).eql('img');
-    await t.expect(firstEmoji.attributes.title).eql(':woman_zombie:');
-    await t.expect(firstEmoji.classNames).eql(['e1']);
-
-    const secondEmoji = await Selector('div.compose img').nth(1)();
-    await t.expect(secondEmoji.tagName).eql('img');
-    await t.expect(secondEmoji.attributes.title).eql(':beers:');
-    await t.expect(secondEmoji.classNames).eql(['e1']);
-});
+    await driver.findElement(By.className('e1-food')).click();
+    await driver.findElement(By.css('.e1._1f37b')).click();
+
+    // Validate emoji
+    const emoji = await driver.findElement(composeArea).findElements(By.xpath('*'));
+    expect(emoji.length).to.equal(2);
+    expect(await emoji[0].getAttribute('title')).to.equal(':woman_zombie:');
+    expect(await emoji[1].getAttribute('title')).to.equal(':beers:');
+
+    // Validate text
+    const html = await driver.findElement(composeArea).getAttribute('innerHTML');
+    expect(/>hi<img/.test(html)).to.be.true;
+}
+
+/**
+ * Insert a newline using shift-enter.
+ */
+async function insertNewline(driver: WebDriver) {
+    // Insert text
+    await driver.findElement(composeArea).click();
+    await driver.findElement(composeArea).sendKeys('hello');
+    await driver.findElement(composeArea).sendKeys(Key.SHIFT, Key.ENTER);
+    await driver.findElement(composeArea).sendKeys('threema');
+    await driver.findElement(composeArea).sendKeys(Key.SHIFT, Key.ENTER);
+    await driver.findElement(composeArea).sendKeys('web');
+
+    const script = `
+        ${extractText.toString()}
+        const element = document.querySelector("div.compose");
+        return extractText(element);
+    `;
+    const text = await driver.executeScript<string>(script);
+    expect(text).to.equal('hello\nthreema\nweb');
+}
+
+// Register tests here
+const TESTS: Array<[string, Testfunc]> = [
+    ['Show and hide emoji selector', showEmojiSelector],
+    ['Insert emoji and text', insertEmoji],
+    ['Insert three lines of text', insertNewline],
+];
+
+// Test runner
+const TEST_URL = 'http://localhost:7777/tests/ui/compose_area.html';
+(async function() {
+    const driver: WebDriver = await new Builder().forBrowser(browser).build();
+    let i = 0;
+    let success = 0;
+    let failed = 0;
+    console.info('\n====== THREEMA WEB UI TESTS ======\n');
+    try {
+        for (const [name, testfunc] of TESTS) {
+            console.info(TermColor.blue(`» ${i + 1}: Running test: ${name}`));
+            await driver.get(TEST_URL);
+            await testfunc(driver);
+            success++;
+            i++;
```

Comment:
I don't see how this will work in case a test fails. IIRC that would break the loop early.

Reply:
Yes, currently the test runner does an early abort. I think that's fine, although we could change the logic so that other tests are run even if there was a failure.